### PR TITLE
feat: add `pk;reproxy`

### PR DESCRIPTION
**Pull request by u1f408**
Adds `pk;reproxy` (`pk;rp`) to reproxy the last sent message, or a linked message (in the same way `pk;edit` does) using the given member, which can be specified as a member name or ID.

This is probably not the best way to do this, very happy to make any changes.

### CODE REVIEW

File: PluralKit.Bot/CommandMeta/CommandHelp.cs [hunk 1]
```diff
@@ -86,6 +86,7 @@ public partial class CommandTree
     public static Command Explain = new Command("explain", "explain", "Explains the basics of systems and proxying");
     public static Command Message = new Command("message", "message <id|link> [delete|author]", "Looks up a proxied message");
     public static Command MessageEdit = new Command("edit", "edit [link] <text>", "Edit a previously proxied message");
+    public static Command MessageReproxy = new Command("reproxy", "reproxy [link] <member>", "Reproxy a previously proxied message using a different member");
```
Comment (spiralw) [inline PluralKit.Bot/CommandMeta/CommandHelp.cs:89]: needs added to docs website

File: PluralKit.Bot/Commands/Message.cs [hunk 2]
```diff
@@ -13,6 +13,8 @@
 
 using NodaTime;
 
+using IMetrics = App.Metrics.IMetrics;
```
Comment (spiralw) [inline PluralKit.Bot/Commands/Message.cs:16]: I think you can just `using App.Metrics;`?

File: PluralKit.Bot/Commands/Message.cs [hunk 3]
```diff
@@ -21,24 +23,98 @@ public class ProxiedMessage
...
+            throw new PKError("Could not reproxy message.");
+
+        // Get a MessageContext for the original message
+        MessageContext? msgCtx =
+               await _repo.GetMessageContext(msg.Message.Sender, msg.Message.Guild!.Value, originalMsg.ChannelId);
+
+        try
+        {
+            var messageChannel = await _cache.GetChannel(originalMsg.ChannelId);
+            var rootChannel = await _cache.GetRootChannel(originalMsg.ChannelId);
```
Comment (spiralw) [inline PluralKit.Bot/Commands/Message.cs:80]: you can't use cache here, clusters don't (currently) share cache - see GetMessage

File: PluralKit.Bot/Commands/Message.cs [hunk 4]
```diff
@@ -21,24 +23,98 @@ public class ProxiedMessage
...
         _rest = rest;
         _webhookExecutor = webhookExecutor;
+        _repo = repo;
         _logChannel = logChannel;
-        // _cache = cache;
+        _cache = cache;
+        _metrics = metrics;
+    }
+
+    public async Task ReproxyMessage(Context ctx)
```
Comment (spiralw) [inline PluralKit.Bot/Commands/Message.cs:50]: make sure this method makes all the necessary permission checks that need to happen when re-proxying a message from DMs / another server. it's possible someone might try to reproxy a message after having send message permissions removed

**Review by spiralw — CHANGES_REQUESTED** [review]
- I would prefer if this code was in ProxyService, a lot of the logic is duplicated / you are missing some important stuff from `HandleProxyExecutedActions`

- should the reproxy command check if the message it's about to reproxy is the last message sent in the channel? it'd be fairly confusing to re-order messages after the fact. maybe a time check (this message was sent x time ago)?

File: PluralKit.Bot/Commands/Message.cs (hunk 2 above)
Comment (u1f408) [reply PluralKit.Bot/Commands/Message.cs:16]: Doing that makes the compiler yell because `App.Metrics` has an incompatible `IClock` definition (which is pulled in by `using NodaTime;`)

**Comment by u1f408** [comment]
Will move into `ProxyService` - wasn't sure where the best place to put this was!

File: PluralKit.Bot/Commands/Message.cs (hunk 2 above)
Comment (spiralw) [reply PluralKit.Bot/Commands/Message.cs:16]: in that case I'd change to use `SystemClock.Instance` instead of the DI IClock, but it's probably irrelevant now since the code's getting moved to ProxyService anyway

File: PluralKit.Bot/Commands/Message.cs [hunk 5]
```diff
@@ -22,23 +24,65 @@ public class ProxiedMessage
...
+        if (target == null)
+            throw new PKError("Could not find a member to reproxy the message with.");
+
+        // Fetch members and get the ProxyMember for `target`
+        List <ProxyMember> members;
+        using (_metrics.Measure.Timer.Time(BotMetrics.ProxyMembersQueryTime))
+            members = (await _repo.GetProxyMembers(ctx.Author.Id, msg.Message.Guild!.Value)).ToList();
+        var match = members.Find(x => x.Id == target.Id);
+        if (match == null)
+            throw new PKError("Could not find a member to reproxy the message with.");
```
Comment (u1f408) [inline PluralKit.Bot/Commands/Message.cs:71]: not sure whether to move this section into the reproxy method in ProxyService, it kinda makes more sense to us to have the member searching where it is currently? (that being said, we're happy to move it if y'all think it makes more sense)

**Comment by u1f408** [comment]
a time check on reproxying is probably a good idea, tbh - i'm hesitant to make it only work if the message was the latest in the channel, as you'd never be able to _use_ the reproxy in faster-moving channels at that point. maybe "can only reproxy if message is less than 1min old" would work?

File: PluralKit.Bot/Proxy/ProxyService.cs [hunk 6]
```diff
@@ -188,6 +188,56 @@ public bool ShouldProxy(Channel channel, Message msg, MessageContext ctx)
...
+            Attachments = originalMsg.Attachments!,
+            FileSizeLimit = guild.FileSizeLimit(),
+            Embeds = originalMsg.Embeds!.ToArray(),
+            Stickers = originalMsg.StickerItems!,
+            AllowEveryone = allowEveryone
+        });
+
+        var sentMessage = await HandleProxyExecutedActions(ctx, autoproxySettings, trigger, proxyMessage, match, deletePrevious: false);
+        await _rest.DeleteMessage(originalMsg.ChannelId!, originalMsg.Id!);
+        await _logChannel.LogMessage(ctx, sentMessage, trigger, proxyMessage, originalMsg.Content!);
```
Comment (spiralw) [inline PluralKit.Bot/Proxy/ProxyService.cs:238]: `HandleProxyExecutedActions` already does all of this, not sure why it's duplicated here

File: PluralKit.Bot/Commands/Message.cs [hunk 7]
```diff
@@ -22,23 +24,70 @@ public class ProxiedMessage
...
+        if (ctx.System.Id != msg.System?.Id)
+            throw new PKError("Can't reproxy a message sent by a different system.");
+
+        // Bail if the message is more than 2 minutes old
+        var msgTimestamp = Instant.FromUnixTimeMilliseconds((long)(msg.Message.Mid >> 22) + 1420070400000);
+        if (msgTimestamp.Plus(Duration.FromSeconds(60 * 2)) < SystemClock.Instance.GetCurrentInstant())
+            throw new PKError("The message is too old to be reproxied.");
+
+        // Get target member ID
+        var target = await ctx.MatchMember();
```
Comment (spiralw) [inline PluralKit.Bot/Commands/Message.cs:65]: I think most people expect this command to work with proxy tags instead of a member name/id

File: PluralKit.Bot/Commands/Message.cs [hunk 8]
```diff
@@ -22,23 +24,70 @@ public class ProxiedMessage
...
+            throw new PKError("Could not find a member to reproxy the message with.");
+
+        try
+        {
+            await _proxy.ExecuteReproxy(ctx.Message, msg.Message, match);
+
+            if (ctx.Guild == null)
+                await _rest.CreateReaction(ctx.Channel.Id, ctx.Message.Id, new Emoji { Name = Emojis.Success });
+            if ((await ctx.BotPermissions).HasFlag(PermissionSet.ManageMessages))
+                await _rest.DeleteMessage(ctx.Channel.Id, ctx.Message.Id);
```
Comment (spiralw) [inline PluralKit.Bot/Commands/Message.cs:85]: this is already done by HandleProxyExecutedActions, you just need to delete the original proxied message

File: PluralKit.Bot/Commands/Message.cs (hunk 5 above)
Comment (spiralw) [reply PluralKit.Bot/Commands/Message.cs:76]: it's probably fine here, yeah

File: PluralKit.Bot/Commands/Message.cs [hunk 9]
```diff
@@ -22,23 +24,70 @@ public class ProxiedMessage
...
+        // Get target member ID
+        var target = await ctx.MatchMember();
+        var previousPtr = ctx.Parameters._ptr;
+        if (target == null)
+            throw new PKError("Could not find a member to reproxy the message with.");
+
+        // Fetch members and get the ProxyMember for `target`
+        List <ProxyMember> members;
+        using (_metrics.Measure.Timer.Time(BotMetrics.ProxyMembersQueryTime))
+            members = (await _repo.GetProxyMembers(ctx.Author.Id, msg.Message.Guild!.Value)).ToList();
```
Comment (spiralw) [inline PluralKit.Bot/Commands/Message.cs:73]: this is fairly expensive, is there a reason why you can't directly use the PKMember from MatchMember?

File: PluralKit.Bot/Commands/Message.cs (hunk 7 above)
Comment (spiralw) [inline PluralKit.Bot/Commands/Message.cs:65]:
```suggestion
        var target = await ctx.MatchMember(restrictToSystem: ctx.System.Id);
```

File: PluralKit.Bot/Commands/Message.cs [hunk 10]
```diff
@@ -22,23 +24,70 @@ public class ProxiedMessage
...
+            throw new PKError("Can't reproxy a message sent by a different system.");
+
+        // Bail if the message is more than 2 minutes old
+        var msgTimestamp = Instant.FromUnixTimeMilliseconds((long)(msg.Message.Mid >> 22) + 1420070400000);
+        if (msgTimestamp.Plus(Duration.FromSeconds(60 * 2)) < SystemClock.Instance.GetCurrentInstant())
+            throw new PKError("The message is too old to be reproxied.");
+
+        // Get target member ID
+        var target = await ctx.MatchMember();
+        var previousPtr = ctx.Parameters._ptr;
```
Comment (spiralw) [inline PluralKit.Bot/Commands/Message.cs:66]:
```suggestion
```

File: PluralKit.Bot/Proxy/ProxyService.cs [hunk 11]
```diff
@@ -188,6 +188,56 @@ public bool ShouldProxy(Channel channel, Message msg, MessageContext ctx)
...
+    {
+        var originalMsg = await _rest.GetMessageOrNull(msg.Channel, msg.Mid);
+        if (originalMsg == null)
+            throw new PKError("Could not reproxy message.");
+
+        // Get a MessageContext for the original message
+        MessageContext ctx =
+            await _repo.GetMessageContext(msg.Sender, msg.Guild!.Value, msg.Channel);
+
+        var autoproxySettings = await _repo.GetAutoproxySettings(ctx.SystemId.Value, msg.Guild!.Value, null);
```
Comment (spiralw) [inline PluralKit.Bot/Proxy/ProxyService.cs:201]: this seems unused?

File: PluralKit.Bot/Proxy/ProxyService.cs [hunk 12]
```diff
@@ -188,6 +188,56 @@ public bool ShouldProxy(Channel channel, Message msg, MessageContext ctx)
...
     }
 
+    public async Task ExecuteReproxy(Message trigger, PKMessage msg, ProxyMember member)
+    {
+        var originalMsg = await _rest.GetMessageOrNull(msg.Channel, msg.Mid);
+        if (originalMsg == null)
+            throw new PKError("Could not reproxy message.");
+
+        // Get a MessageContext for the original message
+        MessageContext ctx =
```
Comment (spiralw) [inline PluralKit.Bot/Proxy/ProxyService.cs:198]: should check `ctx.InBlacklist` and throw early

File: PluralKit.Bot/Proxy/ProxyService.cs [hunk 13]
```diff
@@ -338,6 +389,9 @@ Task SaveMessageInDatabase()
 
         async Task DeleteProxyTriggerMessage()
         {
+            if (!deletePrevious)
```
Comment (spiralw) [inline PluralKit.Bot/Proxy/ProxyService.cs:392]: what is this for?

**Comment by spiralw** [comment]
> maybe "can only reproxy if message is less than 1min old" would work?

sure, that's fine

File: PluralKit.Bot/Proxy/ProxyService.cs (hunk 13 above)
Comment (u1f408) [reply PluralKit.Bot/Proxy/ProxyService.cs:392]: the task in HandleProxyExecuteActions will delete the reproxied message if we delete the trigger, hence some of the duplication of message deletion stuff & this check to return early in that task

File: PluralKit.Bot/Proxy/ProxyService.cs (hunk 11 above)
Comment (u1f408) [reply PluralKit.Bot/Proxy/ProxyService.cs:201]: autoproxySettings is only here because it needs to be passed to HandleProxyExecutedActions, we don't actually need it otherwise

File: PluralKit.Bot/Commands/Message.cs (hunk 9 above)
Comment (u1f408) [reply PluralKit.Bot/Commands/Message.cs:73]: we get a ProxyMember here because that has the utility methods for retrieving server name/avatar/etc - can figure out getting those another way if this is too expensive?

File: PluralKit.Bot/Commands/Message.cs (hunk 7 above)
Comment (u1f408) [reply PluralKit.Bot/Commands/Message.cs:65]: problem with using proxy tags here is that we then have no easy way to distinguish between a prefix and a suffix. if proxy tags were prefix only (or suffix only)  then that would work, but they're not - so you'd have to parse them like the proxy tag set method does (with the `text` sentinel). looking up a member by name/ID makes this a lot easier both in implementation and in usage

File: PluralKit.Bot/Commands/Message.cs (hunk 8 above)
Comment (u1f408) [reply PluralKit.Bot/Commands/Message.cs:85]: this is done here so that this commands acts the same way as `pk;edit` - deleting the trigger if in a channel, marking it as successful with an emoji in DMs. figured it's best to have consistency between the message-editing-related commands?

(that is also why we added the "don't delete the trigger" override flag in HandleProxyExecutedActions)

Comment (spiralw) [reply PluralKit.Bot/Commands/Message.cs:85]: ohh, okay, that makes sense

File: PluralKit.Bot/Commands/Message.cs (hunk 9 above)
Comment (spiralw) [reply PluralKit.Bot/Commands/Message.cs:72]: I looked at the code and it's probably fine like this

File: PluralKit.Bot/Commands/Message.cs (hunk 7 above)
Comment (spiralw) [reply PluralKit.Bot/Commands/Message.cs]: I was going to suggest using a ProxyTagParser then I remembered this method doesn't actually take any content. makes sense to use a member name/id then

File: PluralKit.Bot/Commands/Message.cs (hunk 2 above)
Comment (u1f408) [reply PluralKit.Bot/Commands/Message.cs:16]: taking a quick look at this, removing the DI IClock (and the `NodaTime` import) would require rewriting unrelated stuff around the `pk;edit` timeout, and we don't really wanna mess with that in this PR tbh!

Comment (spiralw) [reply PluralKit.Bot/Commands/Message.cs:16]: oh, that's fine. I'll fix it myself when I merge the PR then

File: PluralKit.Bot/Commands/Message.cs [hunk 14]
```diff
@@ -22,23 +24,69 @@ public class ProxiedMessage
...
+
+    public async Task ReproxyMessage(Context ctx)
+    {
+        var msg = await GetMessageToEdit(ctx);
+
+        if (ctx.System.Id != msg.System?.Id)
+            throw new PKError("Can't reproxy a message sent by a different system.");
+
+        // Bail if the message is more than 2 minutes old
+        var msgTimestamp = Instant.FromUnixTimeMilliseconds((long)(msg.Message.Mid >> 22) + 1420070400000);
```
Comment (spiralw) [inline PluralKit.Bot/Commands/Message.cs:58]: is it possible to move this check inside of `GetMessageToEdit`? maybe add another line next to EditTimeout and pass it into the method

File: PluralKit.Bot/Commands/Message.cs [hunk 15]
```diff
@@ -22,23 +24,69 @@ public class ProxiedMessage
...
+        _repo = repo;
         _logChannel = logChannel;
         // _cache = cache;
+        _metrics = metrics;
+        _proxy =  proxy;
+    }
+
+    public async Task ReproxyMessage(Context ctx)
+    {
+        var msg = await GetMessageToEdit(ctx);
```
Comment (spiralw) [inline PluralKit.Bot/Commands/Message.cs:52]: the error messages in this method all mention editing rather than reproxying

File: PluralKit.Bot/Proxy/ProxyService.cs [hunk 16]
```diff
@@ -307,8 +361,9 @@ private string FixSameNameInner(string name)
     public static bool IsUnlatch(Message message)
         => message.Content.StartsWith(@"\\") || message.Content.StartsWith("\\\u200b\\");
 
-    private async Task HandleProxyExecutedActions(MessageContext ctx, AutoproxySettings autoproxySettings,
-                                                Message triggerMessage, Message proxyMessage, ProxyMatch match)
+    private async Task<PKMessage> HandleProxyExecutedActions(MessageContext ctx, AutoproxySettings autoproxySettings,
```
Comment (spiralw) [inline PluralKit.Bot/Proxy/ProxyService.cs:364]: since `sentMessage` isn't being used upstream, can you revert this change?

**Review by spiralw — CHANGES_REQUESTED** [review]

File: PluralKit.Bot/Commands/Message.cs (hunk 15 above)
Comment (u1f408) [reply PluralKit.Bot/Commands/Message.cs]: not sure what's best to do with the errors in `GetMessageToEdit` - passing in a `bool isReproxy` and changing the messages based on that? 🤷

Comment (spiralw) [reply PluralKit.Bot/Commands/Message.cs]: yeah, that should be fine
